java.security.PrivilegedActionException: java.security.PrivilegedActionException: org.elixir_lang.sdk.erlang_dependent.MissingErlangSdk: Elixir SDK `Elixir 1.13.4` is missing Erlang SDK at java.base/java.security.AccessController.doPrivileged(Native Method)

# System

Plugin Version: 13.1.1
Application: IntelliJ IDEA Ultimate Edition (2022.1.3)
Operating System: Mac OS X (12.3)


# Event


## Exception


### Stacktrace


<details>

<summary>
java.security.PrivilegedActionException: java.security.PrivilegedActionException: org.elixir_lang.sdk.erlang_dependent.MissingErlangSdk: Elixir SDK `Elixir 1.13.4` is missing Erlang SDK
</summary>

```
java.security.PrivilegedActionException: java.security.PrivilegedActionException: org.elixir_lang.sdk.erlang_dependent.MissingErlangSdk: Elixir SDK `Elixir 1.13.4` is missing Erlang SDK
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:748)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:898)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.java:820)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:743)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$6(IdeEventQueue.java:439)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:803)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$7(IdeEventQueue.java:438)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:119)
	at com.intellij.ide.IdeEventQueue.performActivity(IdeEventQueue.java:604)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:436)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:873)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:484)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: java.security.PrivilegedActionException: org.elixir_lang.sdk.erlang_dependent.MissingErlangSdk: Elixir SDK `Elixir 1.13.4` is missing Erlang SDK
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:95)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:751)
	at java.desktop/java.awt.EventQueue$5.run(EventQueue.java:749)
	... 20 more
Caused by: org.elixir_lang.sdk.erlang_dependent.MissingErlangSdk: Elixir SDK `Elixir 1.13.4` is missing Erlang SDK
	at org.elixir_lang.Elixir.elixirSdkToEnsuredErlangSdk(Elixir.kt:35)
	at org.elixir_lang.Elixir.commandLine(Elixir.kt:19)
	at org.elixir_lang.Mix.commandLine(Mix.kt:19)
	at org.elixir_lang.Mix.commandLine$default(Mix.kt:12)
	at org.elixir_lang.new_project_wizard.Step.setupProject(Step.kt:85)
```

</details>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -587,6 +587,9 @@
 * [#3083](https://github.com/KronicDeth/intellij-elixir/pull/3083) - [@KronicDeth](https://github.com/KronicDeth)
   * Ignore `null` operands for `not in` `primaryArguments()`.
     Instead of including both left and right operand always and so having null operand in fixed size array return array with only non-null operands.
+* [#3085](https://github.com/KronicDeth/intellij-elixir/pull/3085) - [@KronicDeth](https://github.com/KronicDeth)
+  * Don't allow Elixir SDK without Erlang SDK to be selected for New Project.
+    The SDK has to be able to work for `mix new`.
 
 ## v14.0.0
 

--- a/resources/META-INF/changelog.html
+++ b/resources/META-INF/changelog.html
@@ -44,6 +44,10 @@
           array with only non-null operands.
         </p>
       </li>
+      <li>
+        <p>Don't allow Elixir SDK without Erlang SDK to be selected for New Project.</p>
+        <p>The SDK has to be able to work for <code>mix new</code>.</p>
+      </li>
     </ul>
   </li>
 </ul>

--- a/src/org/elixir_lang/new_project_wizard/Step.kt
+++ b/src/org/elixir_lang/new_project_wizard/Step.kt
@@ -3,12 +3,17 @@ package org.elixir_lang.new_project_wizard
 import com.intellij.execution.ExecutionException
 import com.intellij.execution.process.ProcessOutput
 import com.intellij.execution.util.ExecUtil
+import com.intellij.ide.JavaUiBundle
+import com.intellij.ide.util.projectWizard.WizardContext
 import com.intellij.ide.wizard.AbstractNewProjectWizardStep
 import com.intellij.ide.wizard.NewProjectWizardBaseData
 import com.intellij.ide.wizard.NewProjectWizardLanguageStep
 import com.intellij.notification.NotificationGroupManager
 import com.intellij.notification.NotificationType
 import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.observable.properties.ObservableMutableProperty
+import com.intellij.openapi.observable.properties.ObservableProperty
+import com.intellij.openapi.options.ConfigurationException
 import com.intellij.openapi.progress.ProgressIndicator
 import com.intellij.openapi.progress.ProgressManager
 import com.intellij.openapi.progress.Task
@@ -16,10 +21,16 @@ import com.intellij.openapi.project.Project
 import com.intellij.openapi.projectRoots.Sdk
 import com.intellij.openapi.projectRoots.SdkTypeId
 import com.intellij.openapi.roots.ProjectRootManager
-import com.intellij.openapi.roots.ui.configuration.sdkComboBox
+import com.intellij.openapi.roots.ui.configuration.*
+import com.intellij.openapi.roots.ui.configuration.projectRoot.ProjectSdksModel
+import com.intellij.openapi.ui.ValidationInfo
+import com.intellij.openapi.util.Disposer
+import com.intellij.openapi.util.NlsContexts
 import com.intellij.ui.dsl.builder.*
 import com.intellij.ui.dsl.gridLayout.HorizontalAlign
+import com.intellij.ui.layout.ValidationInfoBuilder
 import kotlinx.coroutines.CancellationException
+import org.elixir_lang.Elixir
 import org.elixir_lang.Mix
 import org.elixir_lang.module.ElixirModuleBuilder
 import org.elixir_lang.module.ElixirModuleType
@@ -49,8 +60,7 @@ class Step(parent: NewProjectWizardLanguageStep) : AbstractNewProjectWizardStep(
     override fun setupUI(builder: Panel) {
         with(builder) {
             row("Elixir SDK:") {
-                val sdkTypeFilter = { it: SdkTypeId -> it == Type.instance; }
-                sdkComboBox(context, sdkProperty, ElixirModuleType.MODULE_TYPE_ID, sdkTypeFilter)
+                elixirSdkComboBox(context, sdkProperty)
                     .columns(COLUMNS_MEDIUM)
             }
             collapsibleGroup("mix new Options") {
@@ -169,6 +179,64 @@ class Step(parent: NewProjectWizardLanguageStep) : AbstractNewProjectWizardStep(
             Logger.getInstance(javaClass).error(ioException)
 
             throw ioException
+        }
+    }
+}
+
+private val ELIXIR_SDK_TYPE_FILTER = { it: SdkTypeId -> it == Type.instance; }
+private val ANY_SDK_FILTER: ((Sdk) -> Boolean) = { true }
+private val ANY_SDK_TYPE_FILTER: ((SdkTypeId) -> Boolean) = { true }
+private val ANY_SUGGESTED_SDK_FILTER: ((SdkListItem.SuggestedItem) -> Boolean) = { true }
+private val NEW_SDK_CALLBACK_DEFAULT: ((Sdk) -> Unit) = {}
+
+fun Row.elixirSdkComboBox(context: WizardContext, sdkProperty: ObservableMutableProperty<Sdk?>): Cell<JdkComboBox> {
+    val sdksModel = ProjectSdksModel()
+
+    Disposer.register(context.disposable) {
+        sdksModel.disposeUIResources()
+    }
+
+    val comboBox = createSdkComboBox(
+        context.project,
+        sdksModel,
+        sdkProperty,
+        ElixirModuleType.MODULE_TYPE_ID,
+        ELIXIR_SDK_TYPE_FILTER,
+        ANY_SDK_FILTER,
+        ANY_SUGGESTED_SDK_FILTER,
+        ANY_SDK_TYPE_FILTER,
+        NEW_SDK_CALLBACK_DEFAULT
+    )
+
+    return cell(comboBox)
+        .validationOnApply { validateElixirSdk(sdkProperty, sdksModel) }
+        .onApply { context.projectJdk = sdkProperty.get() }
+}
+
+fun ValidationInfoBuilder.validateElixirSdk(
+    sdkProperty: ObservableProperty<Sdk?>,
+    sdkModel: ProjectSdksModel
+): ValidationInfo? = validateAndGetSdkValidationMessage(sdkProperty, sdkModel)?.let { error(it) }
+
+private fun validateAndGetSdkValidationMessage(
+    sdkProperty: ObservableProperty<Sdk?>,
+    sdkModel: ProjectSdksModel
+): @NlsContexts.DialogMessage String? {
+    val sdk = sdkProperty.get()
+
+    return if (sdk == null) {
+        JavaUiBundle.message("title.no.jdk.specified")
+    } else {
+        if (Elixir.elixirSdkHasErlangSdk(sdk)) {
+            try {
+                sdkModel.apply(null, true)
+
+                null
+            } catch (e: ConfigurationException) {
+                e.message ?: e.title
+            }
+        } else {
+            "Internal Erlang SDK is not in Elixir SDK.  Set it before picking this Elixir SDK."
         }
     }
 }

